File: kotlin/bc-ur/src/test/kotlin/com/blockchaincommons/bcur/URCodableTest.kt
package com.blockchaincommons.bcur

import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.Tag
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Tests the URCodable interface.
 *
 * Mirrors the Rust `ur_codable.rs` test which defines a simple `Test` type
 * that implements `CBORTaggedEncodable`/`CBORTaggedDecodable` with tag 24/"leaf",
 * encodes to a UR, and verifies the string and round-trip.
 *
 * In the Rust bc-ur design, `ur()` stores the *untagged* CBOR in the UR.
 * Decoding reads the UR type from the UR type string and reconstructs
 * the object from the untagged CBOR directly via `from_untagged_cbor`.
 */
class URCodableTest {

    /** A simple test type that encodes its string payload as tagged CBOR. */
    private data class TestLeaf(val s: String) : URCodable {
        override fun cborTags(): List<Tag> = listOf(Tag(24, "leaf"))

        override fun untaggedCbor(): Cbor = Cbor.fromString(s)

        companion object {
            fun fromUntaggedCbor(cbor: Cbor): TestLeaf =
                TestLeaf(cbor.tryText())

            fun fromUrString(urString: String): TestLeaf {
                val ur = UR.fromUrString(urString)
                // The UR holds untagged CBOR; decode directly from it.
                return fromUntaggedCbor(ur.cbor)
            }
        }
    }

    @Test
    fun testUrCodable() {
        val test = TestLeaf("test")
        val ur = test.ur()
        val urString = ur.string
        assertEquals("ur:leaf/iejyihjkjygupyltla", urString)

        val test2 = TestLeaf.fromUrString(urString)
        assertEquals(test.s, test2.s)
    }
}

File: typescript/bc-ur/tests/ur-codable.test.ts
import {
  cbor,
  createTag,
  createTaggedCbor,
  extractTaggedContent,
  validateTag,
  type Cbor,
} from "@bc/dcbor";

import {
  fromUrString,
  ur,
  urString,
  type URCodable,
} from "../src/index.js";

class TestValue implements URCodable<TestValue> {
  readonly s: string;

  constructor(s: string) {
    this.s = s;
  }

  cborTags() {
    return [createTag(24, "leaf")];
  }

  untaggedCbor(): Cbor {
    return cbor(this.s);
  }

  taggedCbor(): Cbor {
    return createTaggedCbor(this);
  }

  fromUntaggedCbor(value: Cbor): TestValue {
    return new TestValue(value.toText());
  }

  fromTaggedCbor(value: Cbor): TestValue {
    validateTag(value, this.cborTags());
    return this.fromUntaggedCbor(extractTaggedContent(value));
  }
}

describe("URCodable", () => {
  test("test_ur_codable", () => {
    const value = new TestValue("test");

    const urValue = ur(value);
    expect(urString(value)).toBe("ur:leaf/iejyihjkjygupyltla");
    expect(urValue.toString()).toBe("ur:leaf/iejyihjkjygupyltla");

    const decoded = fromUrString(value, urValue.toString());
    expect(decoded.s).toBe(value.s);
  });
});
